[Unified PDF] [macOS] Can't use the arrow keys to navigate between pages in Single page mode

whsieh opened the pull request:
#### b32dad0669e5ac9124d50dfe8036704e0f6c6aea
<pre>
[Unified PDF] [macOS] Can&apos;t use the arrow keys to navigate between pages in Single page mode
<a href="https://bugs.webkit.org/show_bug.cgi?id=272828">https://bugs.webkit.org/show_bug.cgi?id=272828</a>
<a href="https://rdar.apple.com/123193004">rdar://123193004</a>

Reviewed by Abrar Protyasha.

Implement keyboard scrolling when viewing PDFs in Single Page mode. The behavior (mostly) matches
that in prior macOS versions, but with the added nicety that keyboard scrolling is now smoothly
animated (rather than discrete).

Test: UnifiedPDF.KeyboardScrollingInSinglePageMode

* Source/WebCore/platform/KeyboardScrollingAnimator.h:
* Source/WebCore/platform/ScrollAnimator.h:

`WEBCORE_EXPORT` several helper methods, so we can use them to service keyboard scrolling in
`UnifiedPDFPlugin`.

* Source/WebCore/plugins/PluginViewBase.h:
(WebCore::PluginViewBase::setPDFDisplayModeForTesting):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::setPDFDisplayModeForTesting const):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:

Add an `internals` testing hook to set the current PDF display mode. The arguments are strings that
match the internal `enum class` names:

```
SinglePageDiscrete
SinglePageContinuous
TwoUpDiscrete
TwoUpContinuous
```

* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
(WebKit::PDFPluginBase::annotationRectsForTesting const):
(WebKit::PDFPluginBase::setPDFDisplayModeForTesting):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::didChangeScrollOffset):

See comments in `handleKeyboardEventForDiscreteDisplayMode` below.

(WebKit::UnifiedPDFPlugin::isInDiscreteDisplayMode const):

Rename this to `isInDiscreteDisplayMode`, since it&apos;s now used to guard keyboard scrolling behaviors
that are specific to discrete display modes (both two-up and single).

(WebKit::UnifiedPDFPlugin::isShowingTwoPages const):

Add a helper method that returns whether or not we&apos;re in two-up mode. Currently, only used in a
couple places to determine the offset of the next set of pages when scrolling with arrow keys, if
the page is (/pages are) shorter than the viewport.

(WebKit::UnifiedPDFPlugin::pageBoundsInContentsSpace const):

Pull this out into a helper method, so we can use it from multiple places.

(WebKit::UnifiedPDFPlugin::updateSnapOffsets):

Make a slight adjustment here to snap to the *closest* snap offset rather than the first offset past
the current scroll position. This ensures that the `m_currentlySnappedPage` doesn&apos;t prematurely jump
to the previous page in `handleKeyboardEventForDiscreteDisplayMode` when determining whether we&apos;re
near a snap point.

(WebKit::UnifiedPDFPlugin::determineCurrentlySnappedPage):
(WebKit::UnifiedPDFPlugin::shouldDisplayPage):
(WebKit::UnifiedPDFPlugin::checkedKeyboardScrollingAnimator const):
(WebKit::UnifiedPDFPlugin::handleKeyboardCommand):

Factor this out into a separate method to make the code easier to follow.

(WebKit::UnifiedPDFPlugin::handleKeyboardEventForDiscreteDisplayMode):

Implement the main keyboard scrolling logic here — if the page is shorter than the viewport, we want
to instantly jump to the next or previous page (or set of 2 pages). If the page is taller than the
viewport (e.g. when zooming within a page in Single Page mode), we want to allow the user to scroll
within the page, and clamp scrolling to the edges of the page.

To achieve this, we detect when the user is about to scroll near either the top or bottom page
extent — in this case, we:

1. Stop animated keyboard scrolling immediately.
2. Scroll (with animation) to the edge of the page.

Note that in order for this to work, we also need logic within `didChangeScrollOffset` to also drive
this &quot;snap to edge&quot; animation. This is because we aren&apos;t guaranteed to receive key events in
`handleKeyboardEvent` when the user is about to scroll past the end of the page, if the user has
disabled key repeat in macOS preferences, so we can&apos;t rely (solely) on the logic here to snap to the
edge.

That said, we *additionally* need logic here to prevent key events from starting animated keyboard
scrolling (and snap to the edge instead), otherwise we end up with a stuttering scroll offset if the
user tries to scroll against the edge of the page with arrow keys.

(WebKit::UnifiedPDFPlugin::snapToNearbyPageExtentForKeyboardScrolling):
(WebKit::UnifiedPDFPlugin::animatedScrollDidEnd):
(WebKit::UnifiedPDFPlugin::handleKeyboardEvent):
(WebKit::UnifiedPDFPlugin::performContextMenuAction):
(WebKit::UnifiedPDFPlugin::setPDFDisplayModeForTesting):
(WebKit::UnifiedPDFPlugin::setDisplayModeAndUpdateLayout):
(WebKit::UnifiedPDFPlugin::shouldUseScrollSnapping const): Deleted.
* Source/WebKit/WebProcess/Plugins/PluginView.cpp:
(WebKit::PluginView::setPDFDisplayModeForTesting):
* Source/WebKit/WebProcess/Plugins/PluginView.h:
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/UnifiedPDFTests.mm: Added.
(TestWebKitAPI::sampleColorsInWebView):
(TestWebKitAPI::TEST(UnifiedPDF, KeyboardScrollingInSinglePageMode)):

Add a test to verify that keyboard scrolling works in single page mode (both vertical and horizontal
scrolling), when zoomed in.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIAction.mm:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIMenus.mm:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIWebRequest.mm:

Add missing includes in unrelated API tests, which now cause build failures when adding the new
test source file.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/multiple-pages.pdf: Added.
* Tools/TestWebKitAPI/cocoa/TestWKWebView.h:
* Tools/TestWebKitAPI/cocoa/TestWKWebView.mm:
(-[TestWKWebView sendKey:code:isDown:modifiers:]):
(-[TestWKWebView typeCharacter:modifiers:]):

Split this out into another helper method that only sends keydown or keyup.

Canonical link: <a href="https://commits.webkit.org/277656@main">https://commits.webkit.org/277656@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b4b6a5e371e8fe62b0ab9da81bbe23bd4a75675b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/48241 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/27453 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/51195 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/50929 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/44306 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/50546 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/33387 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/24972 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/39402 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/48823 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/25149 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/41681 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/20550 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/22629 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/42859 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/6297 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/44588 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/43312 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/52833 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/23288 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/19641 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/46741 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/24553 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/41860 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/45649 "Found 1 new API test failure: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/text/replaced-objects (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10640 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/25358 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/24276 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1296,7 +1296,7 @@ static String mutationObserverNotificationString()
         return { };
 
-    auto topLeftInDocumentSpace = convertDown(CoordinateSpace::Plugin, CoordinateSpace::PDFDocumentLayout, FloatPoint { });
+    auto topLeftInDocumentSpace = convertDown(CoordinateSpace::Plugin, CoordinateSpace::Contents, FloatPoint { });
```
What is this change for?

aprotyas replied:
Can you rename the variable to reflect the coordinate space it's in following this change?

whsieh replied:
Oh dear — good catch! That was totally unintended (I was modifying some of this existing code while trying to understand how each of the coordinate spaces work in `UnifiedPDFPlugin`). Will revert this.

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -2212,19 +2230,143 @@ static bool isContextMenuEvent(const WebMouseEvent& event)
+        return false;
+    }
+
+    if (key == "ArrowUp"_s && currentPage) {
+        auto previousPageContentRect = pageBoundsInContentsSpace(currentPage - 1);
+        return scrollAnimator().scrollToPositionWithoutAnimation({ static_cast<float>(m_scrollOffset.width()), previousPageContentRect.y() });
+    }
```
Do we not want the same kind of "skip two pages if necessary" behavior? As you did for ArrowDown

whsieh replied:
So this is only necessary when moving down, since the current page index in 2-up mode will be `0`, `2`, `4`, `6`, … (so going back by 1 page will yield the correct previous offset)

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -3768,6 +3908,8 @@ static NSStringCompareOptions compareOptionsForFindOptions(WebCore::FindOptions
 #endif
 
+
+
```
Remove these extra lines?

whsieh replied:
(Whoops)

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -3790,6 +3932,33 @@ static NSStringCompareOptions compareOptionsForFindOptions(WebCore::FindOptions
 
+void UnifiedPDFPlugin::setPDFDisplayModeForTesting(const String& mode)
+{
+    setDisplayModeAndUpdateLayout([&] {
```
Nit: just capture `mode`?

aprotyas reviewed (approved):
LGTM with one question about respecting single vs double display mode when skipping backwards.

whsieh committed:
[Unified PDF] [macOS] Can't use the arrow keys to navigate between pages in Single page mode
https://bugs.webkit.org/show_bug.cgi?id=272828
rdar://123193004

Reviewed by Abrar Protyasha.

Implement keyboard scrolling when viewing PDFs in Single Page mode. The behavior (mostly) matches
that in prior macOS versions, but with the added nicety that keyboard scrolling is now smoothly
animated (rather than discrete).

Test: UnifiedPDF.KeyboardScrollingInSinglePageMode

* Source/WebCore/platform/KeyboardScrollingAnimator.h:
* Source/WebCore/platform/ScrollAnimator.h:

`WEBCORE_EXPORT` several helper methods, so we can use them to service keyboard scrolling in
`UnifiedPDFPlugin`.

* Source/WebCore/plugins/PluginViewBase.h:
(WebCore::PluginViewBase::setPDFDisplayModeForTesting):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::setPDFDisplayModeForTesting const):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:

Add an `internals` testing hook to set the current PDF display mode. The arguments are strings that
match the internal `enum class` names:

```
SinglePageDiscrete
SinglePageContinuous
TwoUpDiscrete
TwoUpContinuous
```

* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
(WebKit::PDFPluginBase::annotationRectsForTesting const):
(WebKit::PDFPluginBase::setPDFDisplayModeForTesting):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::didChangeScrollOffset):

See comments in `handleKeyboardEventForDiscreteDisplayMode` below.

(WebKit::UnifiedPDFPlugin::isInDiscreteDisplayMode const):

Rename this to `isInDiscreteDisplayMode`, since it's now used to guard keyboard scrolling behaviors
that are specific to discrete display modes (both two-up and single).

(WebKit::UnifiedPDFPlugin::isShowingTwoPages const):

Add a helper method that returns whether or not we're in two-up mode. Currently, only used in a
couple places to determine the offset of the next set of pages when scrolling with arrow keys, if
the page is (/pages are) shorter than the viewport.

(WebKit::UnifiedPDFPlugin::pageBoundsInContentsSpace const):

Pull this out into a helper method, so we can use it from multiple places.

(WebKit::UnifiedPDFPlugin::updateSnapOffsets):

Make a slight adjustment here to snap to the *closest* snap offset rather than the first offset past
the current scroll position. This ensures that the `m_currentlySnappedPage` doesn't prematurely jump
to the previous page in `handleKeyboardEventForDiscreteDisplayMode` when determining whether we're
near a snap point.

(WebKit::UnifiedPDFPlugin::determineCurrentlySnappedPage):
(WebKit::UnifiedPDFPlugin::shouldDisplayPage):
(WebKit::UnifiedPDFPlugin::checkedKeyboardScrollingAnimator const):
(WebKit::UnifiedPDFPlugin::handleKeyboardCommand):

Factor this out into a separate method to make the code easier to follow.

(WebKit::UnifiedPDFPlugin::handleKeyboardEventForDiscreteDisplayMode):

Implement the main keyboard scrolling logic here — if the page is shorter than the viewport, we want
to instantly jump to the next or previous page (or set of 2 pages). If the page is taller than the
viewport (e.g. when zooming within a page in Single Page mode), we want to allow the user to scroll
within the page, and clamp scrolling to the edges of the page.

To achieve this, we detect when the user is about to scroll near either the top or bottom page
extent — in this case, we:

1. Stop animated keyboard scrolling immediately.
2. Scroll (with animation) to the edge of the page.

Note that in order for this to work, we also need logic within `didChangeScrollOffset` to also drive
this "snap to edge" animation. This is because we aren't guaranteed to receive key events in
`handleKeyboardEvent` when the user is about to scroll past the end of the page, if the user has
disabled key repeat in macOS preferences, so we can't rely (solely) on the logic here to snap to the
edge.

That said, we *additionally* need logic here to prevent key events from starting animated keyboard
scrolling (and snap to the edge instead), otherwise we end up with a stuttering scroll offset if the
user tries to scroll against the edge of the page with arrow keys.

(WebKit::UnifiedPDFPlugin::snapToNearbyPageExtentForKeyboardScrolling):
(WebKit::UnifiedPDFPlugin::animatedScrollDidEnd):
(WebKit::UnifiedPDFPlugin::handleKeyboardEvent):
(WebKit::UnifiedPDFPlugin::performContextMenuAction):
(WebKit::UnifiedPDFPlugin::setPDFDisplayModeForTesting):
(WebKit::UnifiedPDFPlugin::setDisplayModeAndUpdateLayout):
(WebKit::UnifiedPDFPlugin::shouldUseScrollSnapping const): Deleted.
* Source/WebKit/WebProcess/Plugins/PluginView.cpp:
(WebKit::PluginView::setPDFDisplayModeForTesting):
* Source/WebKit/WebProcess/Plugins/PluginView.h:
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/UnifiedPDFTests.mm: Added.
(TestWebKitAPI::sampleColorsInWebView):
(TestWebKitAPI::TEST(UnifiedPDF, KeyboardScrollingInSinglePageMode)):

Add a test to verify that keyboard scrolling works in single page mode (both vertical and horizontal
scrolling), when zoomed in.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIAction.mm:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIMenus.mm:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIWebRequest.mm:

Add missing includes in unrelated API tests, which now cause build failures when adding the new
test source file.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/multiple-pages.pdf: Added.
* Tools/TestWebKitAPI/cocoa/TestWKWebView.h:
* Tools/TestWebKitAPI/cocoa/TestWKWebView.mm:
(-[TestWKWebView sendKey:code:isDown:modifiers:]):
(-[TestWKWebView typeCharacter:modifiers:]):

Split this out into another helper method that only sends keydown or keyup.

Canonical link: https://commits.webkit.org/277656@main
